{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment 5: Link Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this assignment, we will use network-based algorithms, such as PageRank, to improve our search results. After the last assignment on evaluating IR systems, we go back to our PubMed dataset of scientific papers. In this dataset, we look at two graphs in particular: the co-authorship network and the citation network.\n",
    "\n",
    "The citation network is similar to the link network of the web: Citations are like web links pointing to other documents. We can therefore apply the same network-based ranking methods."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code from previous assignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle, bz2\n",
    "from collections import defaultdict, namedtuple, Counter\n",
    "from math import log10, sqrt\n",
    "from IPython.display import display, HTML\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# show plots inline within the notebook\n",
    "%matplotlib inline\n",
    "# set plots' resolution\n",
    "plt.rcParams['savefig.dpi'] = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "Ids_file = 'data/allergy_Ids.pkl.bz2'\n",
    "Summaries_file = 'data/allergy_Summaries.pkl.bz2'\n",
    "Citations_file = 'data/allergy_Citations.pkl.bz2'\n",
    "Abstracts_file = 'data/allergy_Abstracts.pkl.bz2'\n",
    "\n",
    "Ids = pickle.load( bz2.BZ2File( Ids_file, 'rb' ) )\n",
    "Summaries = pickle.load( bz2.BZ2File( Summaries_file, 'rb' ) )\n",
    "Citations = pickle.load( bz2.BZ2File( Citations_file, 'rb' ) )\n",
    "Abstracts = pickle.load( bz2.BZ2File( Abstracts_file, 'rb' ) )\n",
    "\n",
    "paper = namedtuple( 'paper', ['title', 'authors', 'year', 'doi'] )\n",
    "\n",
    "for (id, paper_info) in Summaries.items():\n",
    "    Summaries[id] = paper( *paper_info )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def display_summary( id, show_abstract=False, show_id=True, extra_text='' ):\n",
    "    \"\"\"\n",
    "    Function for printing a paper's summary through IPython's Rich Display System.\n",
    "    Trims long author lists, and adds a link to the paper's DOI (when available).\n",
    "    \"\"\"\n",
    "    s = Summaries[id]\n",
    "    lines = []\n",
    "    title = s.title\n",
    "    if s.doi != '':\n",
    "        title = '<a href=http://dx.doi.org/{:s}>{:s}</a>'.format(s.doi, title)\n",
    "    title = '<strong>' + title + '</strong>'\n",
    "    lines.append(title)\n",
    "    authors = ', '.join( s.authors[:20] ) + ('' if len(s.authors) <= 20 else ', ...')\n",
    "    lines.append(str(s.year) + '. ' + authors)\n",
    "    if (show_abstract):\n",
    "        lines.append('<small><strong>Abstract:</strong> <em>{:s}</em></small>'.format(Abstracts[id]))\n",
    "    if (show_id):\n",
    "        lines.append('[ID: {:d}]'.format(id))\n",
    "    if (extra_text != ''):\n",
    "         lines.append(extra_text)\n",
    "    display( HTML('<br>'.join(lines)) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tokenize(text):\n",
    "    return text.split(' ')\n",
    "\n",
    "def preprocess(tokens):\n",
    "    result = []\n",
    "    for token in tokens:\n",
    "        result.append(token.lower())\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "inverted_index = defaultdict(set)\n",
    "\n",
    "for (id, abstract) in Abstracts.items():\n",
    "    for term in preprocess(tokenize(abstract)):\n",
    "        inverted_index[term].add(id)\n",
    "\n",
    "tf_matrix = defaultdict(Counter)\n",
    "\n",
    "for (doc_id, abstract) in Abstracts.items():\n",
    "    tokens = preprocess(tokenize(abstract))\n",
    "    tf_matrix[doc_id] = Counter(tokens)\n",
    "\n",
    "def tf(t,d):\n",
    "    return float(tf_matrix[d][t])\n",
    "\n",
    "def df(t):\n",
    "    return float(len(inverted_index[t]))\n",
    "\n",
    "num_documents = float(len(Abstracts))\n",
    "\n",
    "def idf(t):\n",
    "    return log10((num_documents + 1)/(df(t) + 1))\n",
    "\n",
    "def tfidf(t,d):\n",
    "    return tf(t,d) * idf(t)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Co-authorship network"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We start by building a mapping from authors to the set of identifiers of papers they authored.  We'll be using Python sets again for that purpose."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "papers_of_author = defaultdict(set)\n",
    "\n",
    "for (id, p) in Summaries.items():\n",
    "    for a in p.authors:\n",
    "        papers_of_author[a].add(id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's try it out:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{21436399, 24037376}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "papers_of_author['Chawla A']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1038/nature12526>Type 2 innate lymphoid cells control eosinophil homeostasis.</a></strong><br>2013. Nussbaum JC, Van Dyken SJ, von Moltke J, Cheng LE, Mohapatra A, Molofsky AB, Thornton EE, Krummel MF, Chawla A, Liang HE, Locksley RM<br>[ID: 24037376]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1126/science.1201475>Eosinophils sustain adipose alternatively activated macrophages associated with glucose homeostasis.</a></strong><br>2011. Wu D, Molofsky AB, Liang HE, Ricardo-Gonzalez RR, Jouihan HA, Bando JK, Chawla A, Locksley RM<br>[ID: 21436399]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for id in papers_of_author['Chawla A']:\n",
    "    display_summary(id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now build a co-authorship network, that is a graph linking authors to the set of co-authors they have published with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "coauthors = defaultdict(set)\n",
    "\n",
    "for p in Summaries.values():\n",
    "    for a in p.authors:\n",
    "        coauthors[a].update(p.authors)\n",
    "\n",
    "# The code above results in each author being listed as having co-authored with himself/herself.\n",
    "# We remove these self-references here:\n",
    "for (a, ca) in coauthors.items():\n",
    "    ca.remove(a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And let's try it out again:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Molofsky AB, Van Dyken SJ, Krummel MF, Wu D, Mohapatra A, Nussbaum JC, von Moltke J, Ricardo-Gonzalez RR, Bando JK, Locksley RM, Liang HE, Jouihan HA, Thornton EE, Cheng LE\n"
     ]
    }
   ],
   "source": [
    "print(', '.join( coauthors['Chawla A'] ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Unlike a citation or link network, the edges of this co-authorship network are not directed: There is no direction (no arrow) in the link between author 'Chawla A' and 'Thornton EE', for example. With our chosen implementation, each of these links in fact appears twice in our data, as we also get 'Chawla A' as co-author when we look for 'Thornton EE':"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Molofsky AB, Van Dyken SJ, Krummel MF, Chawla A, Mohapatra A, Nussbaum JC, von Moltke J, Locksley RM, Liang HE, Cheng LE\n"
     ]
    }
   ],
   "source": [
    "print(', '.join( coauthors['Thornton EE'] ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With that in mind, we can calculate some basic statistics about our graph:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes (authors):  137755\n",
      "Number of links (co-authorship relations):  961501\n"
     ]
    }
   ],
   "source": [
    "print('Number of nodes (authors): ', len(coauthors))\n",
    "\n",
    "# We divide by two here to account for the fact that each edge is represented twice (see above):\n",
    "coauthor_rel_count = int(sum( len(c) for c in coauthors.values() ) / 2)\n",
    "print('Number of links (co-authorship relations): ', coauthor_rel_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With this data at hand, we can plot the [degree distribution](https://en.wikipedia.org/wiki/Degree_distribution) by showing the number of collaborators a scientist has published with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZUAAAEKCAYAAADaa8itAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAHLxJREFUeJzt3XuYVdV9//H3J6B4iYpctBRMBuPkgqnxQtU0Nj+jqeLlJ+SpRvwZQwwNTxJviU0rmifBmp958Nf81No2WiJEtEYk5gJJaBBvuTTxMiARlSROhcgEKxgQqVgV/PaPvY4ehjMzmzn7nMOZ83k9z3nO3uusvff3bJj5zt5r7bUUEZiZmRXhLY0OwMzMBg4nFTMzK4yTipmZFcZJxczMCuOkYmZmhXFSMTOzwjipmJlZYZxUzMysME4qZmZWmMGNDqDeRowYEW1tbY0Ow8ysqSxduvT5iBjZV72WSyptbW10dHQ0Ogwzs6Yi6Xd56vn2l5mZFcZJxczMCuOkYmZmhXFSMTOzwjipmJlZYZxUzMysME4qZmZWGCcVMzMrjJOKmZkVpuWeqK+Xtuk/6vGz1TNPq2MkZmb14ysVMzMrjK9UqtDb1YiZWSvylYqZmRXGScXMzArjpGJmZoVxUjEzs8I4qZiZWWFqllQkzZG0TtLjZWV/L+nXkh6T9D1JQ8s+u1xSp6TfSDq5rHxCKuuUNL2sfKykhyQ9JelOSbvX6ruYmVk+tbxSuQWY0K1sCfDeiDgM+C1wOYCkccBk4NC0zdclDZI0CPhn4BRgHHBOqgtwDXBdRLQDG4GpNfwuZmaWQ82SSkT8FNjQrezuiNiaVh8ExqTlicC8iHglIlYBncDR6dUZEU9HxKvAPGCiJAEnAHel7ecCk2r1XczMLJ9GPvz4SeDOtDyaLMmUdKUygDXdyo8BhgMvlCWo8vq7vJ4emvTwLWbW7BrSUC/pi8BW4PZSUYVq0Y/yno43TVKHpI7169fvbLhmZpZT3ZOKpCnA6cC5EVFKBF3AQWXVxgBreyl/HhgqaXC38ooiYlZEjI+I8SNHjizmi5iZ2Q7qmlQkTQAuA86IiC1lHy0EJksaImks0A48DDwCtKeeXruTNeYvTMnofuDMtP0UYEG9voeZmVVWyy7FdwC/BN4lqUvSVOCfgH2AJZKWS7oJICKeAOYDTwI/Bi6IiG2pzeRCYDGwEpif6kKWnC6V1EnWxjK7Vt/FzMzyqVlDfUScU6G4x1/8EXE1cHWF8kXAogrlT5P1DjMzs12En6g3M7PCOKmYmVlhnFTMzKwwTipmZlYYTyecg6cNNjPLx1cqZmZWGCcVMzMrjJOKmZkVxknFzMwK46RiZmaFcVIxM7PCOKmYmVlhnFTMzKwwTipmZlYYJxUzMyuMk4qZmRXGScXMzArjpGJmZoVxUjEzs8I4qZiZWWGcVMzMrDBOKmZmVhgnFTMzK4yTipmZFaZmSUXSHEnrJD1eVjZM0hJJT6X3/VO5JN0gqVPSY5KOLNtmSqr/lKQpZeVHSVqRtrlBkmr1XczMLJ8+k4qkd0gakpaPl3SxpKE59n0LMKFb2XTg3ohoB+5N6wCnAO3pNQ24MR1vGDADOAY4GphRSkSpzrSy7bofy8zM6izPlcp3gG2SDgFmA2OBb/W1UUT8FNjQrXgiMDctzwUmlZXfGpkHgaGSRgEnA0siYkNEbASWABPSZ/tGxC8jIoBby/ZlZmYNkiepvB4RW4GPANdHxOeBUf083oER8SxAej8glY8G1pTV60plvZV3VSivSNI0SR2SOtavX9/P0M3MrC95ksprks4BpgA/TGW7FRxHpfaQ6Ed5RRExKyLGR8T4kSNH9jNEMzPrS56kcj7wfuDqiFglaSzwr/083nPp1hXpfV0q7wIOKqs3BljbR/mYCuVmZtZAg3v7UNIg4IqI+FipLCJWATP7ebyFZFc8M9P7grLyCyXNI2uU3xQRz0paDHy1rHH+JODyiNggabOkY4GHgI8D/9jPmHYZbdN/1ONnq2eeVsdIzMz6p9ekEhHbJI2UtHtEvLozO5Z0B3A8MEJSF1kvrpnAfElTgWeAs1L1RcCpQCewhezqiJQ8vgI8kupdFRGlxv/PkPUw2xP4t/QyM7MG6jWpJKuBf5e0EHipVBgR1/a2UUSc08NHJ1aoG8AFPexnDjCnQnkH8N7eYjAzs/rKk1TWptdbgH1qG46ZmTWzPpNKRPwdgKS9I+KlvuqbmVnryvNE/fslPQmsTOvvk/T1mkdmZmZNJ0+X4uvJnmz/A0BE/Ar4YC2DMjOz5pRrQMmIWNOtaFsNYjEzsyaXp6F+jaQ/A0LS7sDFpFthA0lvz4iYmVk+ea5UPk3W3bc03tbh9ND918zMWlue3l/PA+fWIRYzM2tyfSYVSSOBTwFt5fUj4pO1C8vMzJpRnjaVBcDPgHtwA72ZmfUiT1LZKyIuq3kkZmbW9PI01P9Q0qk1j8TMzJpej1cqkjbz5oRYV0h6BXgtrUdE7FufEM3MrFn0mFQiwoNHmpnZTskz9tdHJO1Xtj5U0qTahmVmZs0oT5vKjIjYVFqJiBfIJtwyMzPbTp6kUqlOnl5jZmbWYvIklQ5J10p6h6SDJV0HLK11YGZm1nzyJJWLgFeBO4H5wMt47C8zM6ug19tYkgYBV0bE39QpHjMza2K9XqlExDbgqDrFYmZmTS5Pg/ujkhYC3wbemKM+Ir5bs6jMzKwp5Ukqw8imEj6hrCwAJxUzM9tOnvlUzi/6oJI+D/wVWXJaAZwPjALmkSWxZcB5EfGqpCHArWS34f4AnB0Rq9N+Lgemko2efHFELC461l1FTzNTrp55Wp0jMTPrWZ75VPYg+8V9KLBHqby/86lIGk02JfG4iHhZ0nxgMnAqcF1EzJN0Uzrmjel9Y0QcImkycA1wtqRxabtDgT8G7pH0ztQOZGZmDZCnS/FtwB8BJwM/AcYAm6s87mBgT0mDgb2AZ8lur92VPp8LlIaCmZjWSZ+fKEmpfF5EvBIRq4BO4Ogq4zIzsyrkSSqHRMSXgJciYi5wGvAn/T1gRPwe+BrwDFky2UT2MOULEbE1VesCRqfl0cCatO3WVH94eXmFbczMrAHyJJXX0vsLkt4L7Ec2tXC/SNqf7CpjLNltq72BUypUjdImPXzWU3mlY06T1CGpY/369TsftJmZ5ZInqcxKieBLwELgSeD/VXHMDwOrImJ9RLxG1ovsz4Ch6XYYZLfY1qblLuAggPT5fsCG8vIK22wnImZFxPiIGD9y5MgqQjczs970mVQi4uaI2BgRP4mIgyPigIi4qYpjPgMcK2mv1DZyIlmiuh84M9WZAixIywvTOunz+yIiUvlkSUMkjQXagYeriMvMzKqUZz6VAyXNlvRvaX2cpKn9PWBEPETW4L6MrDvxW4BZwGXApZI6ydpMZqdNZgPDU/mlwPS0nyfIxiJ7EvgxcIF7fpmZNVaehx9vAb4JfDGt/5ZscMnZPW3Ql4iYwY5zsjxNhd5bEfHfwFk97Odq4Or+xmFmZsXK06YyIiLmA6/DGz2wfEVgZmY7yJNUXpI0nNSzStKxZN16zczMtpPn9telZI3i75D078BI3mxQNzMze0Oesb+WSfpfwLvIng35TeoKbGZmtp08vb/OAvZMva0mAXdKOrLmkZmZWdPJ06bypYjYLOk4svG/5pIN9GhmZradPEml1NPrNODGiFgA7F67kMzMrFnlSSq/l/QvwEeBRWl+kzzbmZlZi8mTHD4KLAYmRMQLZJNo/U1NozIzs6aUZ+yvLcA64LhUtBV4qpZBmZlZc8rT+2sG2bhcl6ei3YB/rWVQZmbWnPLc/voIcAbwEkBErAX2qWVQZmbWnPIklVfTUPOlYVr2rm1IZmbWrPIklfmp99dQSZ8C7gG+UduwzMysGeUZpuVrkv4CeJFsqJYvR8SSmkdmZmZNp9ekImkQsDgiPgw4kZiZWa96vf2VZlLcImm/OsVjZmZNLM/Q9/8NrJC0hNQDDCAiLq5ZVGZm1pTyJJUfpZeZmVmv8jTUz61HIGZm1vyUPYLSOoaMao9RU65vdBh1sXrmaY0OwcwGCElLI2J8X/U82rCZmRWmx6Qi6bb0fkn9wjEzs2bW25XKUZLeDnxS0v6ShpW/6hWgmZk1j96Syk3Aj4F3A0u7vTqqOaikoZLukvRrSSslvT8lqyWSnkrv+6e6knSDpE5Jj0k6smw/U1L9pyRNqSYmMzOrXo9JJSJuiIj3AHMi4uCIGFv2OrjK4/4D8OOIeDfwPmAlMB24NyLagXvTOsApQHt6TQNuBEhXSzOAY4CjgRmlRGRmZo2RZ5Kuz0h6n6QL0+uwag4oaV/gg8DstP9X04ySE4FS9+W5wKS0PBG4NTIPkg1sOQo4GVgSERsiYiPZMDITqonNzMyqk2eSrouB24ED0ut2SRdVccyDgfXANyU9KunmNJz+gRHxLEB6PyDVHw2sKdu+K5X1VF7pO0yT1CGpY9uWTVWEbmZmvcnTpfivgGMi4ssR8WXgWOBTVRxzMHAkcGNEHEE29Mv0XuqrQln0Ur5jYcSsiBgfEeMH7eVhzMzMaiXPMC0CtpWtb6PyL/S8uoCuiHgord9FllSekzQqIp5Nt7fWldU/qGz7McDaVH58t/IHqohrwGmbXnl0HT8UaWa1kudK5ZvAQ5KulHQl8CCpPaQ/IuI/gTWS3pWKTgSeBBYCpR5cU4AFaXkh8PHUC+xYYFO6PbYYOCl1d94fOCmVmZlZg+QZ++taSQ8Ax5FdoZwfEY9WedyLyNpmdgeeBs4nS3DzJU0FngHOSnUXAacCncCWVJeI2CDpK8Ajqd5VEbGhyrjMzKwKeW5/ERHLgGVFHTQilgOVxpA5sULdAC7oYT9zgDlFxWVmZtXx2F9mZlYYJxUzMytMr0lF0iBJ99QrGDMza26eo97MzArjOerNzKwwnqPezMwKk2uOekl7Am+LiN/UISYzM2tSeQaU/N/AcrK5VZB0uKSFtQ7MzMyaT54uxVeSzVfyArzx4OLYGsZkZmZNKk9S2RoR3ceLrzgasJmZtbY8DfWPS/o/wCBJ7cDFwC9qG5aZmTWjPFcqFwGHAq8AdwAvAp+rZVBmZtac8vT+2gJ8UdI12Wpsrn1YZmbWjPL0/vpTSSuAx8gegvyVpKNqH5qZmTWbPG0qs4HPRsTPACQdRzZx12G1DMzMzJpPnjaVzaWEAhARPwd8C8zMzHbQ45WKpCPT4sOS/oWskT6As/Fc8GZmVkFvt7/+f7f1GWXLfk7FzMx20GNSiYgP1TMQMzNrfn021EsaCnwcaCuv76Hvzcysuzy9vxYBDwIrgNdrG46ZmTWzPEllj4i4tOaRmJlZ08vTpfg2SZ+SNErSsNKr5pGZmVnTyXOl8irw98AXebPXVwAHV3NgSYOADuD3EXG6pLHAPGAYsAw4LyJelTQEuBU4CvgDcHZErE77uByYCmwDLo6IxdXE1Crapvc8kefqmafVMRIzG2jyXKlcChwSEW0RMTa9qkooySXAyrL1a4DrIqId2EiWLEjvGyPiEOC6VA9J44DJZINdTgC+nhKVmZk1SJ6k8gSwpciDShoDnAbcnNYFnADclarMBSal5YlpnfT5ian+RGBeRLwSEauATrLJxMzMrEHy3P7aBiyXdD/Z8PdA1V2Krwf+FtgnrQ8HXoiIrWm9CxidlkcDa9Ixt0ralOqPJuuVRoVtzMysAfIkle+nVyEknQ6si4ilko4vFVeoGn181ts23Y85DZgGMGjfkTsVr5mZ5ZdnPpW5fdXZSR8AzpB0KrAHsC/ZlctQSYPT1coYYG2q3wUcBHRJGgzsB2woKy8p36b7d5gFzAIYMqrdQ8yYmdVInvlUVkl6uvurvweMiMsjYkxEtJE1tN8XEecC9wNnpmpTgAVpeWFaJ31+X0REKp8saUjqOdYOPNzfuMzMrHp5bn+NL1veAziLrNtv0S4D5kn6v8CjZPO4kN5vk9RJdoUyGSAinpA0H3gS2ApcEBHbahBXS+mpu7G7GptZHsr+6N/JjaSfR8RxNYin5oaMao9RU65vdBhNx0nFrLVJWhoR4/uql2dAySPLVt9CduWyTw/VzcysheW5/VU+r8pWYDXw0ZpEY2ZmTS1P7y/Pq2JmZrnkuf01BPhLdpxP5arahWVmZs0oz+2vBcAmYCllT9SbmZl1lyepjImICTWPxMzMml6eASV/IelPah6JmZk1vTxXKscBn5C0iuz2l4CIiMNqGpmZmTWdPEnllJpHYWZmA0KeLsW/q0cgZmbW/PK0qZiZmeWS5/aXmee1N7NcfKViZmaFcVIxM7PCOKmYmVlhnFTMzKwwTipmZlYYJxUzMyuMuxRb1dzd2MxKfKViZmaFcVIxM7PCOKmYmVlhnFTMzKwwbqi3muqpEd8N+GYDU92vVCQdJOl+SSslPSHpklQ+TNISSU+l9/1TuSTdIKlT0mOSjizb15RU/ylJU+r9XczMbHuNuP21FfjriHgPcCxwgaRxwHTg3ohoB+5N65BNEtaeXtOAGyFLQsAM4BjgaGBGKRGZmVlj1D2pRMSzEbEsLW8GVgKjgYnA3FRtLjApLU8Ebo3Mg8BQSaOAk4ElEbEhIjYCS4AJdfwqZmbWTUMb6iW1AUcADwEHRsSzkCUe4IBUbTSwpmyzrlTWU3ml40yT1CGpY9uWTUV+BTMzK9OwhnpJbwW+A3wuIl6U1GPVCmXRS/mOhRGzgFkAQ0a1V6xj9eWn8M0GpoZcqUjajSyh3B4R303Fz6XbWqT3dam8CziobPMxwNpeys3MrEEa0ftLwGxgZURcW/bRQqDUg2sKsKCs/OOpF9ixwKZ0e2wxcJKk/VMD/UmpzMzMGqQRt78+AJwHrJC0PJVdAcwE5kuaCjwDnJU+WwScCnQCW4DzASJig6SvAI+keldFxIb6fAWrJT/bYta86p5UIuLnVG4PATixQv0ALuhhX3OAOcVFZ2Zm1fAwLWZmVhgnFTMzK4zH/rKm4W7IZrs+JxUbENy4b7Zr8O0vMzMrjJOKmZkVxre/bEBzO4xZfflKxczMCuOkYmZmhXFSMTOzwrhNxVqW21vMiuekYlaBn3sx6x/f/jIzs8I4qZiZWWF8+8tsJ/TWDtMT3zKzVuKkYlZjTkTWSnz7y8zMCuMrFbNdUH+7O7vXmjWak4pZi3MisiI5qZg1mf600fRnG7P+cFIxs8J4lAJTRDQ6hroaMqo9Rk25vtFhmFmVnKTqS9LSiBjfVz1fqZhZU2p0W5Cvyipr+qQiaQLwD8Ag4OaImNngkMysgdx+1FhNnVQkDQL+GfgLoAt4RNLCiHiysZGZWStr5cTW7A8/Hg10RsTTEfEqMA+Y2OCYzMxaVrMnldHAmrL1rlRmZmYN0NS3vwBVKNuhO5ukacC0tPrK7645/fGaRrXrGwE83+ggdgE+DxmfB5+Dkt7Ow9vz7KDZk0oXcFDZ+hhgbfdKETELmAUgqSNPt7iBzOcg4/OQ8XnwOSgp4jw0++2vR4B2SWMl7Q5MBhY2OCYzs5bV1FcqEbFV0oXAYrIuxXMi4okGh2Vm1rKaOqkARMQiYNFObDKrVrE0EZ+DjM9DxufB56Ck6vPQcsO0mJlZ7TR7m4qZme1CWiapSJog6TeSOiVNb3Q89SJpjqR1kh4vKxsmaYmkp9L7/o2MsdYkHSTpfkkrJT0h6ZJU3mrnYQ9JD0v6VToPf5fKx0p6KJ2HO1OnlwFN0iBJj0r6YVpvxXOwWtIKScsldaSyqn8mWiKplA3ncgowDjhH0rjGRlU3twATupVNB+6NiHbg3rQ+kG0F/joi3gMcC1yQ/v1b7Ty8ApwQEe8DDgcmSDoWuAa4Lp2HjcDUBsZYL5cAK8vWW/EcAHwoIg4v60Zc9c9ESyQVWng4l4j4KbChW/FEYG5angtMqmtQdRYRz0bEsrS8meyXyWha7zxERPxXWt0tvQI4AbgrlQ/48yBpDHAacHNaFy12DnpR9c9EqyQVD+eyvQMj4lnIfuECBzQ4nrqR1AYcATxEC56HdNtnObAOWAL8B/BCRGxNVVrhZ+N64G+B19P6cFrvHED2B8XdkpamUUeggJ+Jpu9SnFOu4VxsYJP0VuA7wOci4sXsD9TWEhHbgMMlDQW+B7ynUrX6RlU/kk4H1kXEUknHl4orVB2w56DMByJiraQDgCWSfl3ETlvlSiXXcC4t5DlJowDS+7oGx1NzknYjSyi3R8R3U3HLnYeSiHgBeICsjWmopNIfmAP9Z+MDwBmSVpPdBj+B7Mqllc4BABGxNr2vI/sD42gK+JlolaTi4Vy2txCYkpanAAsaGEvNpXvms4GVEXFt2Uetdh5GpisUJO0JfJisfel+4MxUbUCfh4i4PCLGREQb2e+B+yLiXFroHABI2lvSPqVl4CTgcQr4mWiZhx8lnUr2F0lpOJerGxxSXUi6AziebPTR54AZwPeB+cDbgGeAsyKie2P+gCHpOOBnwArevI9+BVm7Siudh8PIGl8Hkf1BOT8irpJ0MNlf7cOAR4GPRcQrjYu0PtLtry9ExOmtdg7S9/1eWh0MfCsirpY0nCp/JlomqZiZWe21yu0vMzOrAycVMzMrjJOKmZkVxknFzMwK46RiZmaFcVIxKyPpAUk1n6tc0sVp1OTba32sPuL4hKQ/LltfLWlEI2Oy5tYqw7SY1ZykwWXjR/Xls8ApEbGqljHl8Amyh96qfoJ8J7+/DVC+UrGmI6kt/ZX/jTQvyN3pCfHtrjQkjUjDcZT+Iv++pB9IWiXpQkmXpjk1HpQ0rOwQH5P0C0mPSzo6bb+3srlpHknbTCzb77cl/QC4u0Ksl6b9PC7pc6nsJuBgYKGkz3erP0jS19I8F49JuiiVn5iOuyLFMaTCsd4q6V5Jy1K9Uoxt2n4+nS9IulLSmcB44PY0p8aeqcpFZft4d9pmWDp/j6XzdVgqv1LSLEl3A7dKOlTZnC3LU932nfrHteYXEX751VQvoI1sjpTD0/p8siegIRvPanxaHgGsTsufADqBfYCRwCbg0+mz68gGmSxt/420/EHg8bT81bJjDAV+C+yd9tsFDKsQ51FkT/HvDbwVeAI4In22GhhRYZvPkI1RNjitDwP2IBtl+52p7NZSvN22HQzsW/bdO8kGS2wrfY/02ReAK7ufr7K4LkrLnwVuTsv/CMxIyycAy9PylcBSYM+yeuem5d1L5X61zstXKtasVkXE8rS8lOwXZ1/uj4jNEbGeLKn8IJWv6Lb9HfDGXDT7pvGyTgKmp2HjHyD7Rf+2VH9JVB7K4jjgexHxUmTzmHwX+PM+YvwwcFOk20hpv+9K3/e3qc5csoTXnYCvSnoMuIds+PYD+zheJaUBN8vP63HAbSmm+4DhkvZLny2MiJfT8i+BKyRdBry9rNxahJOKNavycZm28Wb74Fbe/H+9Ry/bvF62/jrbty92H7soyH5h/2Vks+QdHhFvi4jSzIEv9RBjf8bWV4XjV9yPpGPSbablks4AziW7CjsqIg4nG+ttD7Y/J7DjeemudF7Kz2tvw8O/8f0j4lvAGcDLwGJJJ/RxLBtgnFRsoFlNdtsJ3hx1dmedDW8MRLkpIjYBi8naGpQ+OyLHfn4KTJK0VxoJ9iNkA1v25m7g06Vh2FNbz6+BNkmHpDrnAT+JiIfKktxCYD+yuUJek/Qh4O2p/nPAAZKGp7aY08uOt5nslmCe73Juiul44PmIeLF7pTRQ4dMRcQPZiLeH5di3DSDu/WUDzdeA+ZLOA+7r5z42SvoFsC/wyVT2FbJRrh9LiWU12/9y3kFELJN0C/BwKro5Ih7t49g3A+9Mx3mNrH3nnySdD3w7JZtHgJsqbHs78ANJHcBysmRESjJXkY3KvKpUntwC3CTpZeD9vcR1JfDNdGttC28Oj97d2WQdHV4D/hO4qo/vawOMRyk2M7PC+PaXmZkVxknFzMwK46RiZmaFcVIxM7PCOKmYmVlhnFTMzKwwTipmZlYYJxUzMyvM/wCY2dBPEljIDwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist( x=[ len(ca) for ca in coauthors.values() ], bins=range(60) )\n",
    "plt.xlabel('number of co-authors')\n",
    "plt.ylabel('number of researchers')\n",
    "plt.xlim(0,51);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have explored this network, let's move to the citation network."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Citations network"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we can look at the citation network. In contrast to the co-authorship network, the citation network is a _directed_ network, where edges can be drawn as arrows. We'll start by expanding the our data about citations into two mappings: \n",
    "\n",
    "* `papers_citing[id]`: papers citing a given paper\n",
    "* `cited_by[id]`: papers cited by a given paper (in other words: its list of references)\n",
    "\n",
    "`papers_citing` will give us the list of a node's incoming links, whereas `cited_by` will give us the list of its outgoing links."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "papers_citing = Citations  # no changes needed, this is what we are storing already in the Citations dataset\n",
    "\n",
    "cited_by = defaultdict(list)\n",
    "\n",
    "for ref, papers_citing_ref in papers_citing.items():\n",
    "    for id in papers_citing_ref:\n",
    "        cited_by[ id ].append( ref )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we are dealing with a subset of the data (the papers mentioning \"allergy\"), `papers_citing` can contain references to papers outside of our subset. On the other hand, the way we created `cited_by`, it will only contain backward references from within our dataset, meaning that it is incomplete with respect to the whole dataset. Nethertheless, we can use this citation network on our subset of allergy-related papers to implement link analysis techniques.\n",
    "\n",
    "Let us now look at an exemplary paper, let's say the one with identifier 23917667. We can now use the `cited_by` mapping to retrieve its (incomplete) list of references:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12 references found for paper 23917667\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{21771119: 'The role of genetics and environment in the rise of childhood food allergy.',\n",
       " 21624746: 'Association between genes encoding components of the IL-10/IL-0 receptor pathway and asthma in children.',\n",
       " 18440625: 'Dust mite exposure modifies the effect of functional IL10 polymorphisms on allergy and asthma exacerbations.',\n",
       " 17935766: \"The natural history of IgE-mediated cow's milk allergy.\",\n",
       " 17121586: 'Influence of SNPs in cytokine-related genes on the severity of food allergy and atopic eczema in children.',\n",
       " 16373958: \"Cow's milk allergy: a complex disorder.\",\n",
       " 16310526: 'Adverse reactions to foods.',\n",
       " 15864084: 'Genetic and environmental risk factors for the development of food allergy.',\n",
       " 15131577: \"Cow's milk-specific T-cell reactivity of children with and without persistent cow's milk allergy: key role for IL-10.\",\n",
       " 14748015: 'IL10 gene polymorphisms are associated with asthma phenotypes in children.',\n",
       " 12884306: 'Reduced transforming growth factor-beta1-producing T cells in the duodenal mucosa of children with food allergy.',\n",
       " 12592300: '9. Food allergy.'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paper_id = 23917667\n",
    "refs = { id : Summaries[id].title for id in cited_by[paper_id] }\n",
    "print(len(refs), 'references found for paper', paper_id)\n",
    "refs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we lookup the same paper in `papers_citing`, we now see that some of the cited papers are themselves in our dataset, but others are not (shown below as `'??'`):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{27070332: 'Are genetic tests informative in predicting food allergy?',\n",
       " 26516307: '??',\n",
       " 25721048: '??',\n",
       " 25238536: '??'}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{ id : Summaries.get(id,['??'])[0]  for id in papers_citing[paper_id] }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Paper 26516307, for example, is not in our dataset and we do not have any direct information about it, but its repeated occurrence in other papers' citation lists does allow us to reconstruct some of its references. Below is the list of papers in our dataset cited by that paper:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4 references identified for the paper with id 26516307\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{23917667: \"Interleukin 10 (IL10) and transforming growth factor β1 (TGFβ1) gene polymorphisms in persistent IgE-mediated cow's milk allergy.\",\n",
       " 23108822: 'Interleukin-13 gene polymorphisms at -1055 C/T and +2044 G/A positions in patients with squamous cell carcinoma of head and neck.',\n",
       " 15711639: 'IL-13 R130Q, a common variant associated with allergy and asthma, enhances effector mechanisms essential for human allergic inflammation.',\n",
       " 15548720: 'Reduced immunoglobulin E and allergy among adults with glioma compared with controls.'}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paper_id2 = 26516307\n",
    "refs2 = { id : Summaries[id].title for id in cited_by[paper_id2] }\n",
    "print(len(refs2), 'references identified for the paper with id', paper_id2)\n",
    "refs2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have a better understanding about the data we're dealing with, let us obtain again some basic statistics about our graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of papers in our subset: 73740 (100.00 %)\n",
      "Number of papers cited at least once: 37449 (50.79 %)\n",
      "Number of isolated nodes:  34435 (46.70 %)\n"
     ]
    }
   ],
   "source": [
    "n = len(Ids)\n",
    "print('Number of papers in our subset: {:d} ({:.2f} %)'.format(n, 100.0) )\n",
    "\n",
    "with_citation = [ id for id in Ids if papers_citing[id] != [] ]\n",
    "with_citation_rel = 100. * len(with_citation) / n\n",
    "print('Number of papers cited at least once: {:d} ({:.2f} %)'.format(len(with_citation), with_citation_rel) )\n",
    "\n",
    "isolated = set( id for id in Ids if papers_citing[id] == [] and id not in cited_by )\n",
    "isolated_rel = 100. * len(isolated) / n\n",
    "print('Number of isolated nodes:  {:d} ({:.2f} %)'.format(len(isolated), isolated_rel) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overall number of nodes: 198687 (100.00 %)\n",
      "Number of non-isolated nodes: 164252 (82.67 %)\n",
      "Number of nodes outside our subset: 124947 (62.89 %)\n"
     ]
    }
   ],
   "source": [
    "id_set = set( Ids )\n",
    "citing_set = set( cited_by.keys() )\n",
    "\n",
    "outsiders = citing_set - id_set   # set difference\n",
    "nodes = citing_set | id_set   # set union\n",
    "non_isolated = nodes - isolated   # set difference\n",
    "\n",
    "print('Overall number of nodes: {:d} ({:.2f} %)'.format(len(nodes), 100.0) )\n",
    "\n",
    "non_isolated_rel = 100. * len(non_isolated) / len(nodes)\n",
    "print('Number of non-isolated nodes: {:d} ({:.2f} %)'.format(len(non_isolated), non_isolated_rel) )\n",
    "\n",
    "outsiders_rel = 100. * len(outsiders) / len(nodes)\n",
    "print('Number of nodes outside our subset: {:d} ({:.2f} %)'.format( len(outsiders), outsiders_rel ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overall number of links (citations): 309385 (100.00 %)\n",
      "Citations outside the subset: 221506 (71.60 %)\n"
     ]
    }
   ],
   "source": [
    "all_citations = [ c for citing in papers_citing.values() for c in citing ]\n",
    "outsider_citations = [ c for citing in papers_citing.values() for c in citing if c in outsiders ]\n",
    "\n",
    "print('Overall number of links (citations): {:d} ({:.2f} %)'.format(len(all_citations), 100.0) )\n",
    "\n",
    "outsider_citations_rel = 100. * len(outsider_citations) / len(all_citations)\n",
    "print('Citations outside the subset: {:d} ({:.2f} %)'.format(len(outsider_citations), outsider_citations_rel) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us now find our which 10 papers are the most cited in our dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1016/j.cell.2006.02.015>Pathogen recognition and innate immunity.</a></strong><br>2006. Akira S, Uematsu S, Takeuchi O<br>[ID: 16497588]<br>Citation count: 2655"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1056/NEJMoa1105243>Prevention of HIV-1 infection with early antiretroviral therapy.</a></strong><br>2011. Cohen MS, Chen YQ, McCauley M, Gamble T, Hosseinipour MC, Kumarasamy N, Hakim JG, Kumwenda J, Grinsztejn B, Pilotto JH, Godbole SV, Mehendale S, Chariyalertsak S, Santos BR, Mayer KH, Hoffman IF, Eshleman SH, Piwowar-Manning E, Wang L, Makhema J, ...<br>[ID: 21767103]<br>Citation count: 2519"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1038/35047123>A Toll-like receptor recognizes bacterial DNA.</a></strong><br>2000. Hemmi H, Takeuchi O, Kawai T, Kaisho T, Sato S, Sanjo H, Matsumoto M, Hoshino K, Wagner H, Takeda K, Akira S<br>[ID: 11130078]<br>Citation count: 1357"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1038/nri3073>Protective and pathogenic functions of macrophage subsets.</a></strong><br>2011. Murray PJ, Wynn TA<br>[ID: 21997792]<br>Citation count: 1073"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1038/nri2515>The gut microbiota shapes intestinal immune responses during health and disease.</a></strong><br>2009. Round JL, Mazmanian SK<br>[ID: 19343057]<br>Citation count: 1049"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1126/science.1198469>Induction of colonic regulatory T cells by indigenous Clostridium species.</a></strong><br>2011. Atarashi K, Tanoue T, Shima T, Imaoka A, Kuwahara T, Momose Y, Cheng G, Yamasaki S, Saito T, Ohba Y, Taniguchi T, Takeda K, Hori S, Ivanov II, Umesaki Y, Itoh K, Honda K<br>[ID: 21205640]<br>Citation count: 916"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1086/588660>Revised definitions of invasive fungal disease from the European Organization for Research and Treatment of Cancer/Invasive Fungal Infections Cooperative Group and the National Institute of Allergy and Infectious Diseases Mycoses Study Group (EORTC/MSG) Consensus Group.</a></strong><br>2008. De Pauw B, Walsh TJ, Donnelly JP, Stevens DA, Edwards JE, Calandra T, Pappas PG, Maertens J, Lortholary O, Kauffman CA, Denning DW, Patterson TF, Maschmeyer G, Bille J, Dismukes WE, Herbrecht R, Hope WW, Kibbler CC, Kullberg BJ, Marr KA, ...<br>[ID: 18462102]<br>Citation count: 866"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1146/annurev.immunol.21.120601.141122>Naturally arising CD4+ regulatory t cells for immunologic self-tolerance and negative control of immune responses.</a></strong><br>2004. Sakaguchi S<br>[ID: 15032588]<br>Citation count: 800"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1146/annurev.immunol.25.022106.141623>Regulatory T cells: mechanisms of differentiation and function.</a></strong><br>2012. Josefowicz SZ, Lu LF, Rudensky AY<br>[ID: 22224781]<br>Citation count: 736"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1056/NEJMoa1506816>Initiation of Antiretroviral Therapy in Early Asymptomatic HIV Infection.</a></strong><br>2015. INSIGHT START Study Group., Lundgren JD, Babiker AG, Gordin F, Emery S, Grund B, Sharma S, Avihingsanon A, Cooper DA, Fätkenheuer G, Llibre JM, Molina JM, Munderi P, Schechter M, Wood R, Klingman KL, Collins S, Lane HC, Phillips AN, Neaton JD<br>[ID: 26192873]<br>Citation count: 685"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "citation_count_per_paper = [ (id, len(citations)) for (id,citations) in papers_citing.items() ]\n",
    "sorted_by_citation_count = sorted(citation_count_per_paper, key=lambda i:i[1], reverse=True)\n",
    "\n",
    "for (id, c) in sorted_by_citation_count[:10]:\n",
    "    display_summary(id, extra_text = 'Citation count: ' + str(c))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before we can start calculating some interesting network metrics, we will first have a closer look at the Python package that we are going to use for that."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Link Analysis for Search Engines"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to use the citation network, we need to be able to perform some complex graph algorithms on it. To make our lives easier, we will use [NetworkX](https://pypi.python.org/pypi/networkx), a Python package for dealing with complex networks:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: networkx in /usr/lib/python3/dist-packages (2.2)\r\n"
     ]
    }
   ],
   "source": [
    "! pip install --user networkx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "\n",
    "G = nx.DiGraph(cited_by)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now have a NetworkX Directed Graph stored in `G`, where a node represents a paper, and an edge represents a citation. This means we can now apply the [algorithms](http://networkx.github.io/documentation/networkx-1.10/reference/algorithms.html) and [functions](http://networkx.github.io/documentation/networkx-1.10/reference/functions.html) of NetworkX to our graph:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: \n",
      "Type: DiGraph\n",
      "Number of nodes: 164252\n",
      "Number of edges: 309385\n",
      "Average in degree:   1.8836\n",
      "Average out degree:   1.8836\n",
      "Directed graph: True\n",
      "Density of graph: 1.14678120125452e-05\n"
     ]
    }
   ],
   "source": [
    "print(nx.info(G))\n",
    "print('Directed graph:', nx.is_directed(G))\n",
    "print('Density of graph:', nx.density(G))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As this graph was generated from citations only, we need to add all isolated nodes (nodes that are not cited and do not cite other papers) as well:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "G.add_nodes_from(isolated)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And now we get slightly different values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: \n",
      "Type: DiGraph\n",
      "Number of nodes: 198687\n",
      "Number of edges: 309385\n",
      "Average in degree:   1.5571\n",
      "Average out degree:   1.5571\n",
      "Directed graph: True\n",
      "Density of graph: 7.837228966726296e-06\n"
     ]
    }
   ],
   "source": [
    "print(nx.info(G))\n",
    "print('Directed graph:', nx.is_directed(G))\n",
    "print('Density of graph:', nx.density(G))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we are ready to use this package for our tasks."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tasks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Your name:** ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 1\n",
    "\n",
    "Plot the in-degree distribution (that is, the distribution of the number of incoming links; see [here](https://en.wikipedia.org/wiki/Degree_distribution) and [here](http://mathinsight.org/degree_distribution) for more detailed explanations) for the citation network. What can you tell about the shape of this distribution, and what does this tell us about the network?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Answer:** [_Write your answer text here_]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 2\n",
    "\n",
    "Using the [Link Analysis](https://networkx.github.io/documentation/networkx-1.10/reference/algorithms.link_analysis.html) algorithms provided by NetworkX, calculate the PageRank score for all nodes in the citation network, and store the result in a variable called `pageranks`. Print out the PageRank values for the two example papers given below.\n",
    "You can also use the `pagerank_scipy` implementation, which tends to be considerably faster than its regular `pagerank` counterpart, for which you have to install the [SciPy package](http://scipy.org/):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! pip install --user scipy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To print and compare PageRank values, you might want to use commands like `print('{:.6f}'.format(var))` to use regular decimal notation with a fixed number of decimal places."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add your code here:\n",
    "#pageranks = ...\n",
    "\n",
    "# print PageRank for paper 26968958\n",
    "# print PageRank for paper 20923692"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 3\n",
    "\n",
    "Why do the two papers above have such different PageRank values? Write code below to investigate and show the cause of this, and then explain the cause of this difference based on the results generated by your code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "**Answer:** [_Write your answer text here_]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 4\n",
    "\n",
    "Copy the scoring function `score_ntn` from Task 4 of assignment 3. Rename it to `score_ntn_pagerank` and change its code to incorporate a paper's PageRank score in it's final score, in addition to tf-idf. In other words, the new function should accept a `list` of query tokens and a document ID, and should return a single `float` value that is calculated based on both scores (PageRank and tf-idf). Note that a `tf-idf` function is already provided above. Explain your decision on how to combine the two scores."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "**Answer:** [_Write your answer text here_]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 5\n",
    "\n",
    "Copy the query function `query_ntn` from Task 4 of assignment 3. Rename it to `query_ntn_pagerank` and change the code to use our new scoring function `score_ntn_pagerank` from task 4 above. Demonstrate these functions with an example query that returns paper 19343057 as the top result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Submission"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Submit the answers to the assignment via Canvas as a modified version of this Notebook file (file with `.ipynb` extension) that includes your code and your answers.\n",
    "\n",
    "Before submitting, restart the kernel and re-run the complete code (**Kernel > Restart & Run All**), and then check whether your assignment code still works as expected.\n",
    "\n",
    "Don't forget to add your name, and remember that the assignments have to be done individually and group submissions are **not allowed**."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5rc1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
